TestPods_AddEphemeralContainer became flaky

This test seem to have a bit of a flaky behavior, e.g:

https://github.com/grafana/xk6-disruptor/actions/runs/5772040184/job/15646605631?pr=271

and

https://github.com/grafana/xk6-disruptor/actions/runs/5772040184/job/15646650382?pr=271

I have reproduced this locally on `main` as well, although it may take four or five runs to make it fail:
<details>
<summary>Long log</summary>

```
roobre@Archiroo  ±main ● ☸ kind-e2e-pod-disruptor
20:15:30 ~/Devel/xk6-disruptor $> go clean -testcache && make test
go test -race  ./...
?   	github.com/grafana/xk6-disruptor/cmd/agent	[no test files]
?   	github.com/grafana/xk6-disruptor/cmd/agent/commands	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/agent/protocol	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/internal/version	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/kubernetes	[no test files]
ok  	github.com/grafana/xk6-disruptor	0.076s
?   	github.com/grafana/xk6-disruptor/pkg/runtime/profiler	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/assertions	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/checks	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/cluster	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/deploy	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/fetch	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/fixtures	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubectl	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubernetes	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/grpc	[no test files]
ok  	github.com/grafana/xk6-disruptor/pkg/agent	5.031s
ok  	github.com/grafana/xk6-disruptor/pkg/agent/protocol/grpc	1.027s
ok  	github.com/grafana/xk6-disruptor/pkg/agent/protocol/http	0.031s
ok  	github.com/grafana/xk6-disruptor/pkg/api	0.087s
ok  	github.com/grafana/xk6-disruptor/pkg/disruptors	0.075s
ok  	github.com/grafana/xk6-disruptor/pkg/iptables	0.031s
ok  	github.com/grafana/xk6-disruptor/pkg/kubernetes/helpers	5.084s
ok  	github.com/grafana/xk6-disruptor/pkg/runtime	0.027s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/cluster	0.023s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/command	0.024s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubernetes/namespace	0.039s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/grpc/dynamic	0.029s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/grpc/ping	0.029s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/kubernetes/builders	3.047s
ok  	github.com/grafana/xk6-disruptor/pkg/utils	5.051s

roobre@Archiroo  ±main ● ☸ kind-e2e-pod-disruptor
20:15:40 ~/Devel/xk6-disruptor $> go clean -testcache && make test
go test -race  ./...
?   	github.com/grafana/xk6-disruptor/cmd/agent	[no test files]
?   	github.com/grafana/xk6-disruptor/cmd/agent/commands	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/agent/protocol	[no test files]
ok  	github.com/grafana/xk6-disruptor	0.073s
?   	github.com/grafana/xk6-disruptor/pkg/internal/version	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/kubernetes	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/runtime/profiler	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/assertions	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/checks	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/cluster	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/deploy	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/fetch	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/fixtures	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubectl	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubernetes	[no test files]
?   	github.com/grafana/xk6-disruptor/pkg/testutils/grpc	[no test files]
ok  	github.com/grafana/xk6-disruptor/pkg/agent	5.029s
ok  	github.com/grafana/xk6-disruptor/pkg/agent/protocol/grpc	1.028s
ok  	github.com/grafana/xk6-disruptor/pkg/agent/protocol/http	0.029s
ok  	github.com/grafana/xk6-disruptor/pkg/api	0.092s
ok  	github.com/grafana/xk6-disruptor/pkg/disruptors	0.079s
ok  	github.com/grafana/xk6-disruptor/pkg/iptables	0.038s
--- FAIL: TestPods_AddEphemeralContainer (0.00s)
    --- FAIL: TestPods_AddEphemeralContainer/Create_ephemeral_container_waiting (5.02s)
        pods_test.go:223: unexpected error: ephemeral container for pod "test-pod" has not started after 5.000000s
FAIL
FAIL	github.com/grafana/xk6-disruptor/pkg/kubernetes/helpers	5.065s
ok  	github.com/grafana/xk6-disruptor/pkg/runtime	0.033s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/cluster	0.026s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/command	0.026s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/e2e/kubernetes/namespace	0.039s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/grpc/dynamic	0.029s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/grpc/ping	0.029s
ok  	github.com/grafana/xk6-disruptor/pkg/testutils/kubernetes/builders	3.050s
ok  	github.com/grafana/xk6-disruptor/pkg/utils	5.052s
FAIL
make: *** [Makefile:50: test] Error 1

```

</details>


I initally thought increasing the timeout would help, but still fails now and then even with 10 seconds:

```
--- FAIL: TestPods_AddEphemeralContainer (0.00s)
    --- FAIL: TestPods_AddEphemeralContainer/Create_ephemeral_container_waiting (10.02s)
        pods_test.go:223: unexpected error: ephemeral container for pod "test-pod" has not started after 10.000000s
FAIL
FAIL	github.com/grafana/xk6-disruptor/pkg/kubernetes/helpers	10.076s
```

I suspect of some minor race condition either on the test or on the observer logic.

--- a/pkg/kubernetes/helpers/pods.go
+++ b/pkg/kubernetes/helpers/pods.go
@@ -91,9 +91,25 @@ func (h *podHelper) waitForCondition(
 		},
 	)
 	if err != nil {
-		return false, err
+		return false, fmt.Errorf("starting pod watcher: %w", err)
 	}
 	defer watcher.Stop()
+
+	// we check if the pod already satisfies the condition to prevent race conditions
+	// on which we miss the update that makes the condition true
+	pod, err := h.client.CoreV1().Pods(namespace).Get(
+		ctx,
+		name,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		return false, fmt.Errorf("getting pod: %w", err)
+	}
+
+	condition, err := checker(pod)
+	if condition || err != nil {
+		return condition, err
+	}
 
 	expired := time.After(timeout)
 	for {
@@ -109,7 +125,7 @@ func (h *podHelper) waitForCondition(
 				if !isPod {
 					return false, errors.New("received unknown object while watching for pods")
 				}
-				condition, err := checker(pod)
+				condition, err = checker(pod)
 				if condition || err != nil {
 					return condition, err
 				}

--- a/pkg/kubernetes/helpers/pods_test.go
+++ b/pkg/kubernetes/helpers/pods_test.go
@@ -117,14 +117,12 @@ func TestPods_Wait(t *testing.T) {
 }
 
 func TestPods_AddEphemeralContainer(t *testing.T) {
-	t.Skip("Skipping due to flaky behavior. See issue #280")
 	t.Parallel()
 
 	type TestCase struct {
 		test        string
 		podName     string
 		expectError bool
-		container   string
 		status      corev1.ContainerStatus
 		options     AttachOptions
 	}
@@ -135,7 +133,6 @@ func TestPods_AddEphemeralContainer(t *testing.T) {
 			test:        "Create ephemeral container not waiting",
 			podName:     "test-pod",
 			expectError: false,
-			container:   "ephemeral",
 			status: corev1.ContainerStatus{
 				Name: "ephemeral",
 				State: corev1.ContainerState{
@@ -196,6 +193,7 @@ func TestPods_AddEphemeralContainer(t *testing.T) {
 				pod.Status.EphemeralContainerStatuses = []corev1.ContainerStatus{
 					tc.status,
 				}
+
 				// update pod and stop watching updates
 				return pod, false, nil
 			}
@@ -221,7 +219,7 @@ func TestPods_AddEphemeralContainer(t *testing.T) {
 				tc.options,
 			)
 			if !tc.expectError && err != nil {
-				t.Errorf("unexpected error: %v", err)
+				t.Errorf("failed: %v", err)
 				return
 			}
 		})

